servlet/src/main/scala/org/http4s/servlet/BlockingHttp4sServlet.scala

```diff
@@ -0,0 +1,162 @@
+package org.http4s
+package servlet
+
+import java.net.InetSocketAddress
+import javax.servlet._
+import javax.servlet.http.{HttpServlet, HttpServletRequest, HttpServletResponse, HttpSession}
+
+import cats.effect._
+import cats.implicits.{catsSyntaxEither => _, _}
+import org.http4s.headers.`Transfer-Encoding`
+import org.http4s.server._
+import org.log4s.getLogger
+
+import scala.collection.JavaConverters._
+
+class BlockingHttp4sServlet[F[_]](
+    service: HttpRoutes[F],
+    servletIo: BlockingServletIo[F],
+    serviceErrorHandler: ServiceErrorHandler[F])(implicit F: Effect[F])
+    extends HttpServlet {
+  private[this] val logger = getLogger
+
+  private[this] var serverSoftware: ServerSoftware = _
+
+  // micro-optimization: unwrap the service and call its .run directly
+  private[this] val serviceFn = service.run
+
+  object ServletRequestKeys {
+    val HttpSession: AttributeKey[Option[HttpSession]] = AttributeKey[Option[HttpSession]]
+  }
+
+  override def init(config: ServletConfig): Unit = {
+    val servletContext = config.getServletContext
+    val servletApiVersion = ServletApiVersion(servletContext)
+    logger.info(s"Detected Servlet API version $servletApiVersion")
+
+    serverSoftware = ServerSoftware(servletContext.getServerInfo)
+  }
+
+  override def service(
+      servletRequest: HttpServletRequest,
+      servletResponse: HttpServletResponse): Unit =
+    try {
+      val bodyWriter = servletIo.initWriter(servletResponse)
+
+      val render = toRequest(servletRequest).fold(
+        onParseFailure(_, servletResponse, bodyWriter),
+        handleRequest(_, servletResponse, bodyWriter)
+      )
+
+      F.runAsync(render) {
+          case Right(_) => Sync[IO].unit
+          case Left(t) => Sync[IO].delay(errorHandler(servletResponse)(t))
+        }
+        .unsafeRunSync()
+    } catch errorHandler(servletResponse)
+
+  private def onParseFailure(
```

Comment:
I think everything here below is a copy-paste and could move into an abstract class?  Make them `protected`.